gcp_variant_transforms/libs/generate_report.py

```diff
@@ -0,0 +1,127 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generates conflicts report."""
+
+import csv
+import os
+from typing import Dict, List, Union  # pylint: disable=unused-import
+
+from gcp_variant_transforms.beam_io.vcf_header_io import VcfHeader  # pylint: disable=unused-import
+from gcp_variant_transforms.beam_io.vcf_header_io import VcfParserHeaderKeyConstants
+from gcp_variant_transforms.transforms.merge_header_definitions import Definition  # pylint: disable=unused-import
+from gcp_variant_transforms.transforms.merge_header_definitions import VcfHeaderDefinitions  # pylint: disable=unused-import
+
+
+class ConflictsReporter(object):
+  """Generates the header conflicts report."""
+
+  _HEADER_LINE = ['ID', 'Conflicts', 'Proposed Solution']
+  _REPORT_NAME = 'header_conflicts_report.csv'
+  _NO_CONFLICTS_MESSAGE = ['No conflicts found.']
+  _NO_SOLUTION_MESSAGE = 'Not resolved'
+
+  def __init__(self, directory=''):
+    # type: (str) -> None
+    """Initializes ``ConflictsReporter`` object.
+
+    Args:
+      directory: The location where the conflicts report is saved.
+    """
+    self._directory = directory
+
+  def report_conflicts(self, header_definitions, resolved_header=None):
+    # type: (VcfHeaderDefinitions, VcfHeader) -> None
+    """Generates a report.
+
+    Combines the conflicts extracted from ``header_definitions`` and their
+    resolutions from ``resolved_header`` to generate a conflicts report.
+    """
+    resolved_header = resolved_header or VcfHeader()
+    format_conflicts = self._extract_conflicts(header_definitions.formats)
+    info_conflicts = self._extract_conflicts(header_definitions.infos)
+    format_headers = resolved_header.formats
+    info_headers = resolved_header.infos
+    contents = list()
+    contents.extend(self._generate_contents(format_conflicts, format_headers))
+    contents.extend(self._generate_contents(info_conflicts, info_headers))
+    self._write_to_report(contents)
+
+  def _extract_conflicts(
+      self,
+      definitions  # type: Dict[str, Dict[Definition, List[str]]]
+      ):
+    # type: (...) -> Dict[str, Dict[Definition, List[str]]]
+    """Extracts the fields that have conflicted definitions."""
+    # len(v) > 1 means there are conflicted definitions for this field.
+    return dict([(k, v) for k, v in definitions.items() if len(v) > 1])
+
+  def _generate_contents(
+      self,
+      conflicts,  # type: Dict[str, Dict[Definition, List[str]]]
+      headers  # type: Dict[str, Dict[str, Union[str, int]]
+      ):
+    # type: (...) -> List[str]
+    """Generates the report contents.
```

Comment:
Please give an example of what a typical output looks like (it can be a merge of the two examples below :) ).